### PR TITLE
Working with code reviews at Distancify

### PR DESCRIPTION
**Pull request by skjohansen**
This is the outcome of the discussion we had on Monday the 4th of march on code reviews.
We did never make it to the discussion on coding conventions, and that is the reason for this section to be empty, but please provide input.

### CODE REVIEW

File: development/guidelines/codereview.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+At Distancify are code review (as part of pull request) an integrated part of the daily life of a developer.
+
+The main purpose of our code reviews is to do findings which can't be caught by automated tools.
+
+Stuff we are looking for in out code review is:
+
+- Is the code covered with unit tests? "This requires a unit test"
+- Is the code clear and readable? Naming and semantics, both the reviewer and developer should be able to read the code
+- Do we follow our [coding conventions](coding-conventions.md)? Not all parts can be found by a linting tool
```
Comment (DistantRikard) [inline development/guidelines/codereview.md]: Maybe rephrase to mean something like cover all parts of coding conventions that currently isn't being covered by the linting tool, and aspire to gradually cover more ground with the linting tool as we find time to improve it, and also, evaluate if any exceptions to the coding conventions invoked in the target PR is valid or not

File: development/guidelines/codereview.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
```
Comment (DistantRikard) [inline development/guidelines/codereview.md]: I would also add a line for architectural concerns

File: development/guidelines/codereview.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Code review and pull request
+
+At Distancify are code review (as part of pull request) an integrated part of the daily life of a developer.
+
+The main purpose of our code reviews is to do findings which can't be caught by automated tools.
+
+Stuff we are looking for in out code review is:
+
+- Is the code covered with unit tests? "This requires a unit test"
+- Is the code clear and readable? Naming and semantics, both the reviewer and developer should be able to read the code
```
Comment (DistantRikard) [inline development/guidelines/codereview.md]: I agree with this, don't know if the phrasing can be improved to further empathize something like "maybe you can have some low-level hackish ninjutsu shit pushed down to the bottom layer of your code that isn't necessarily explicitly understood by the reviewer in any other way than that it works because there is a covering unit test that passes, *BUT*, the top layer of your code should be close to readable English much like a script that explains what this code actually does"

Comment (kristofferlindvall) [reply development/guidelines/codereview.md]: The semantics should be clear in relation to the current module, regardless of it's level withing the full stack. You should be able to understand what a module does without having to fully understand why or for who it does it.

File: development/guidelines/codereview.md (hunk 1 above)
Comment (kristofferlindvall) [reply development/guidelines/codereview.md]: To make a reviewers life easier, perhaps we should define what to look for rather that what to not look for? What convention is it that can't be covered by the linting?

File: development/guidelines/codereview.md (hunk 2 above)
Comment (kristofferlindvall) [reply development/guidelines/codereview.md]: What does that mean? I don't want to see code reviews degrading into a never ending cycle of subjective opinions about architecture. For example, we could add that we should look for known anti-patterns and whether APIs are used as originally intended. Other than that there are an almost infinite number of ways to do anything.

File: development/guidelines/codereview.md (hunk 1 above)
Comment (skjohansen) [reply development/guidelines/codereview.md]: First of all we should define our coding conventions in order to configure the linting, and when we know what our coding conventions are should we maybe rephrase this.
My suggestion is that we delete this row for now, and start working on a coding convention.

File: development/guidelines/codereview.md (hunk 2 above)
Comment (skjohansen) [reply development/guidelines/codereview.md]: I agree with @kristofferlindvall, the codereview is not for architectural concerns, but adding known anti-patterns and usage of APIs sound good to me. @DistantRikard ?

File: development/guidelines/codereview.md (hunk 3 above)
Comment (skjohansen) [reply development/guidelines/codereview.md]: @DistantRikard I assume that you refer to legacy code? Otherwise should the low-level hackish ninjutsu shit have been caught in a pull request relevant to that code.

**Comment by lindblom** [comment]
I think we should add something about "approved with suggestions" that i like to use a lot. It should be used when you feel that the things you have noted should be taken into consideration for the future but the developer doesn't necessarily have to go in and update his/her code.

File: development/guidelines/codereview.md (hunk 3 above)
Comment (DistantRikard) [reply development/guidelines/codereview.md]: My explanation of what I wanted to say was bad, I agree/understand what you both said. My comment was coming from a place of classes and functions often ending up being too large/encompassing, and while "clear and readable" probably includes that, I thought maybe this recurring problem could get a explicit mention.

File: development/guidelines/codereview.md (hunk 2 above)
Comment (DistantRikard) [reply development/guidelines/codereview.md]: Not sure if architectural is exactly what I meant, but often, there are comments about the way things are structured. These comments can probably be divided into 2 categories, let's call the first one "according to my experience/taste/inclination, this way of slicing the pie isn't good", and this can indeed lead to 

>code reviews degrading into a never ending cycle of subjective opinions

But on the other hand, sometimes there's useful advice, for example something like "if you do this, and a PBI like this comes up, which it's likely to do, then we will have effectively painted ourselves into a corner", and that type of advice I'm sure has saved us time, and sometimes that has included (but isn't limited to) things like

>whether APIs are used as originally intended

I'm probably being too ambiguous, and the reason for that is probably that I don't know what the dividing line between these two types comments are, I just know that when it comes to the "structural" comments, I definitely feel like there has been both the "subjective bickering" and the "life-saving" kinds, but I obviously agree that

>look for known anti-patterns and whether APIs are used as originally intended

is a good thing.

Comment (skjohansen) [reply development/guidelines/codereview.md]: Sound like this could be "solved" by encourage to constructive criticism in the pull requests.
So if you are about to write that something is wrong you should also provide a reason and in best of all worlds even a suggestion to a better solution. And if you don't, you should be prepared to be ignored.

Comment (kristofferlindvall) [reply development/guidelines/codereview.md]: I agree with Stephan, we can't get stuck in a situation where we think something is wrong but don't have a better idea.
